### ISSUE TITLE
Compile failure, possibly glibc related.

### ISSUE THREAD

**Issue by FrostKnight** [deathkiller/jazz2-native#55]
### Jazz² Resurrection version

2.6.1

### System information

Linux Hyperbola 0.4.4  Intel® HD Graphics 4000

### Issue description

After the make part, it fails at 100% like the below:

/usr/bin/ld: _deps/ncine_libraries-src/Linux/x86_64/libopenmpt.so: undefined reference to `std::__throw_bad_array_new_length()@GLIBCXX_3.4.29'
/usr/bin/ld: _deps/ncine_libraries-src/Linux/x86_64/libopenmpt.so: undefined reference to `__libc_single_threaded@GLIBC_2.32'
/usr/bin/ld: _deps/ncine_libraries-src/Linux/x86_64/libopenmpt.so: undefined reference to `std::to_chars(char*, char*, double)@GLIBCXX_3.4.29'
/usr/bin/ld: _deps/ncine_libraries-src/Linux/x86_64/libopenmpt.so: undefined reference to `std::__glibcxx_assert_fail(char const*, int, char const*, char const*)@GLIBCXX_3.4.30'
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/jazz2.dir/build.make:4688: jazz2] Error 1
make[1]: *** [CMakeFiles/Makefile2:100: CMakeFiles/jazz2.dir/all] Error 2
make: *** [Makefile:156: all] Error 2

### Steps to reproduce

attempted compiling after,

mkdir build in the main folder, cd build, cmake .. CMAKE_SYSTEM_PREFIX_PATH=/usr

then make

at 100% it fails for some reason. No idea why. Maybe glibc related? Sorry, would build from earlier release probably but you have no archives.

I will say, I look forward to online mode.

**Comment by deathkiller** [comment]
The game automatically uses bundled library version if the library (package) is not installed in OS. Because Linux (unlike Windows) is so scattered, the bundled libraries can be sometimes incompatible with OS and other libraries.

You have to install system package for `libopenmpt` library (or somehow get a compatible version of it for your OS).

**Comment by FrostKnight** [comment]
> The game automatically uses bundled library version if the library (package) is not installed in OS. Because Linux (unlike Windows) is so scattered, the bundled libraries can be sometimes incompatible with OS and other libraries.
> 
> You have to install system package for `libopenmpt` library (or somehow get a compatible version of it for your OS).

odd thing is, even after installing that it still is failing. Perhaps my version is too old?

**Comment by deathkiller** [comment]
libopenmpt version usually doesn't matter, because the game uses only old API which is in any version. 
Check CMake log, you should see which libopenmpt library was used for build.

**Comment by FrostKnight** [comment]
> libopenmpt version usually doesn't matter, because the game uses only old API which is in any version. Check CMake log, you should see which libopenmpt library was used for build.

Ahh okay. Well its version 0.5.12 supposedly of libopenmpt

**Comment by FrostKnight** [comment]
[CMakeError.log](https://github.com/deathkiller/jazz2-native/files/15000738/CMakeError.log)

**Comment by deathkiller** [comment]
I meant the CMake output usually written to terminal, it should look like this:
```
-- The CXX compiler identification is GNU 11.4.0
-- The C compiler identification is GNU 11.4.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Compiling for architecture: x86_64
-- Content directory: /home/runner/work/jazz2-native/jazz2-native/Content
-- Found Git: /usr/bin/git (found version "2.43.2")
-- Game version: 2.6.r663-c69758b
-- Downloading dependencies from "https://github.com/deathkiller/jazz2-libraries/archive/2.6.1.tar.gz"...
-- Found ZLIB: /usr/lib/x86_64-linux-gnu/libz.so (found version "1.2.11")
-- Found OpenGL: /usr/lib/x86_64-linux-gnu/libOpenGL.so
-- Found GLFW: /home/runner/work/jazz2-native/jazz2-native/_build/_deps/ncine_libraries-src/Linux/x86_64/libglfw.so
-- Found SDL2: /home/runner/work/jazz2-native/jazz2-native/_build/_deps/ncine_libraries-src/Linux/x86_64/libSDL2main.a;/home/runner/work/jazz2-native/jazz2-native/_build/_deps/ncine_libraries-src/Linux/x86_64/libSDL2.so
-- Found OpenAL: /home/runner/work/jazz2-native/jazz2-native/_build/_deps/ncine_libraries-src/Linux/x86_64/libopenal.so
-- Found Vorbis: /usr/lib/x86_64-linux-gnu/libogg.so
-- Found libopenmpt: /usr/lib/x86_64-linux-gnu/libopenmpt.so
-- Using GLFW as the preferred backend
-- Using GNU IFUNC for CPU-dependent functionality
```

And `Found libopenmpt` should point to your system library. You can also try to delete your build output directory to reset all cached paths.

**Comment by FrostKnight** [comment]
> I meant the CMake output usually written to terminal, it should look like this:
> 
> ```
> -- The CXX compiler identification is GNU 11.4.0
> -- The C compiler identification is GNU 11.4.0
> -- Detecting CXX compiler ABI info
> -- Detecting CXX compiler ABI info - done
> -- Check for working CXX compiler: /usr/bin/g++ - skipped
> -- Detecting CXX compile features
> -- Detecting CXX compile features - done
> -- Detecting C compiler ABI info
> -- Detecting C compiler ABI info - done
> -- Check for working C compiler: /usr/bin/gcc - skipped
> -- Detecting C compile features
> -- Detecting C compile features - done
> -- Compiling for architecture: x86_64
> -- Content directory: /home/runner/work/jazz2-native/jazz2-native/Content
> -- Found Git: /usr/bin/git (found version "2.43.2")
> -- Game version: 2.6.r663-c69758b
> -- Downloading dependencies from "https://github.com/deathkiller/jazz2-libraries/archive/2.6.1.tar.gz"...
> -- Found ZLIB: /usr/lib/x86_64-linux-gnu/libz.so (found version "1.2.11")
> -- Found OpenGL: /usr/lib/x86_64-linux-gnu/libOpenGL.so
> -- Found GLFW: /home/runner/work/jazz2-native/jazz2-native/_build/_deps/ncine_libraries-src/Linux/x86_64/libglfw.so
> -- Found SDL2: /home/runner/work/jazz2-native/jazz2-native/_build/_deps/ncine_libraries-src/Linux/x86_64/libSDL2main.a;/home/runner/work/jazz2-native/jazz2-native/_build/_deps/ncine_libraries-src/Linux/x86_64/libSDL2.so
> -- Found OpenAL: /home/runner/work/jazz2-native/jazz2-native/_build/_deps/ncine_libraries-src/Linux/x86_64/libopenal.so
> -- Found Vorbis: /usr/lib/x86_64-linux-gnu/libogg.so
> -- Found libopenmpt: /usr/lib/x86_64-linux-gnu/libopenmpt.so
> -- Using GLFW as the preferred backend
> -- Using GNU IFUNC for CPU-dependent functionality
> ```
> 
> And `Found libopenmpt` should point to your system library. You can also try to delete your build output directory to reset all cached paths.

I'll rebuild it just to be safe. Maybe doing cmake again will fix it if I do ir from scratch.

**Comment by FrostKnight** [comment]
[output.txt](https://github.com/deathkiller/jazz2-native/files/15000932/output.txt)
You mean like that?

**Comment by deathkiller** [comment]
No, this is build output, but CMake has another output. You should see the same as I shown you 2 messages above.

**Comment by FrostKnight** [comment]
> No, this is build output, but CMake has another output. You should see the same as I shown you 2 messages above.

Oh sorry, I am a bit confused which file?

**Comment by FrostKnight** [comment]
[cmakeerror.txt](https://github.com/deathkiller/jazz2-native/files/15001116/cmakeerror.txt)
You mean this? If not, tell me specifically.

**Comment by deathkiller** [comment]
Yes, the last one, but it's too short. Try to delete everything, redownload it and rerun it, so everything is recreated.

**Comment by deathkiller** [comment]
It looks fine. It doesn't build now?

**Comment by FrostKnight** [comment]
> It looks fine. It doesn't build now?

Yeah... I dunno why.

**Comment by FrostKnight** [comment]
I have the dependencies here: https://aur.archlinux.org/packaes/jazz2-bin

Also, but I wondered if there are any hidden ones I am unaware of

**Comment by FrostKnight** [comment]
Update! 2.6 version works apparently. Just not 2.6.1? I think?

I tried to backtrack to 2.6 instead of the master and it worked.

**Comment by deathkiller** [comment]
I did so many changes between 2.5.0 and master, so it's hard to tell what's different and why it doesn't work for you now. Probably the only way is to find last working commit.

**Comment by deathkiller** [comment]
Added `NCINE_COMPILE_OPENMPT` CMake option to compile libopenmpt from sources instead of using library. It's not tested, but can help with this issue.

**Comment by FrostKnight** [comment]
> Added `NCINE_COMPILE_OPENMPT` CMake option to compile libopenmpt from sources instead of using library. It's not tested, but can help with this issue.

How would I use that? 

I tried adding cmake options, but it was ignored. I am not sure why.

Do I need to put something like, -DCMAKE behind to do this? 

like this:

cmake ..  -DCMAKE CMAKE_PREFIX_PATH=/usr

without the -DCMAKE it fails, but it might still fail regardless, hence my curiosity

**Comment by deathkiller** [comment]
Why do you want to changed the prefix?

You should be able to configure and build it like this:
```
cmake -B output_dir -D CMAKE_BUILD_TYPE=Release -D NCINE_WITH_GLEW=OFF -D NCINE_COMPILE_OPENMPT=OFF
make -C output_dir
```

**Comment by FrostKnight** [comment]
I change the prefix more because that's usually how I do it in Hyperbola.

/usr is where stuff goes not /ust/local

That's how I have been doing it more than anything. 

So you are saying I need to do this in cmake and make for it to work though? Is this correct?

**Comment by deathkiller** [comment]
Yeah, I don't know anything about Hyperbola, but the game builds just fine in Fedora and Ubuntu.

**Comment by deathkiller** [comment]
Closing due to inactivity. Feel free to re-open if needed.

**Comment by FrostKnight** [comment]
> Closing due to inactivity. Feel free to re-open if needed.

Every so often I will probably try to recompile, like right now.

It seems I cannot build it again. I wanted to test your netplay function. ;)

When I type make --debug I get this:

GNU Make 4.3
Built for x86_64-pc-linux-gnu
Copyright (C) 1988-2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Reading makefiles...
Updating makefiles....
Updating goal targets....
 File 'default_target' does not exist.
   File 'all' does not exist.
     File 'cmake_check_build_system' does not exist.
    Must remake target 'cmake_check_build_system'.
    Successfully remade target file 'cmake_check_build_system'.
  Must remake target 'all'.
GNU Make 4.3
Built for x86_64-pc-linux-gnu
Copyright (C) 1988-2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Reading makefiles...
Updating makefiles....
Updating goal targets....
 File 'all' does not exist.
   File 'CMakeFiles/Backward.dir/all' does not exist.
  Must remake target 'CMakeFiles/Backward.dir/all'.
GNU Make 4.3
Built for x86_64-pc-linux-gnu
Copyright (C) 1988-2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Reading makefiles...
Updating makefiles....
Updating goal targets....
 File 'CMakeFiles/Backward.dir/depend' does not exist.
Must remake target 'CMakeFiles/Backward.dir/depend'.
Successfully remade target file 'CMakeFiles/Backward.dir/depend'.
GNU Make 4.3
Built for x86_64-pc-linux-gnu
Copyright (C) 1988-2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Reading makefiles...
Updating makefiles....
Updating goal targets....
 File 'CMakeFiles/Backward.dir/build' does not exist.
   File 'Backward' does not exist.
  Must remake target 'Backward'.
  Successfully remade target file 'Backward'.
Must remake target 'CMakeFiles/Backward.dir/build'.
Successfully remade target file 'CMakeFiles/Backward.dir/build'.
[  0%] Built target Backward
  Successfully remade target file 'CMakeFiles/Backward.dir/all'.
   File 'CMakeFiles/jazz2.dir/all' does not exist.
  Must remake target 'CMakeFiles/jazz2.dir/all'.
GNU Make 4.3
Built for x86_64-pc-linux-gnu
Copyright (C) 1988-2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Reading makefiles...
Updating makefiles....
Updating goal targets....
 File 'CMakeFiles/jazz2.dir/depend' does not exist.
Must remake target 'CMakeFiles/jazz2.dir/depend'.
Successfully remade target file 'CMakeFiles/jazz2.dir/depend'.
GNU Make 4.3
Built for x86_64-pc-linux-gnu
Copyright (C) 1988-2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Reading makefiles...
Updating makefiles....
Updating goal targets....
 File 'CMakeFiles/jazz2.dir/build' does not exist.
   File 'jazz2' does not exist.
     File 'CMakeFiles/jazz2.dir/Sources/Shared/Core/Logger.cpp.o' does not exist.
    Must remake target 'CMakeFiles/jazz2.dir/Sources/Shared/Core/Logger.cpp.o'.
[  0%] Building CXX object CMakeFiles/jazz2.dir/Sources/Shared/Core/Logger.cpp.o

**Comment by FrostKnight** [comment]
> Closing due to inactivity. Feel free to re-open if needed.

Btw, can't reopen.

**Comment by deathkiller** [comment]
There is missing `-D` before `CMAKE_PREFIX_PATH`, isn't it? But I can't see any error there, is the log complete? Also can you enclose the log in triple backticks (```) at the beginning and the end, so it's easier to read, please?

**Comment by FrostKnight** [comment]
> There is missing `-D` before `CMAKE_PREFIX_PATH`, isn't it? But I can't see any error there, is the log complete? Also can you enclose the log in triple backticks (```) at the beginning and the end, so it's easier to read, please?

So I should do it like this instead right?

cmake .. -B output_dir -D CMAKE_BUILD_TYPE=Release -D NCINE_WITH_GLEW=OFF -D NCINE_COMPILE_OPENMPT=OFF -DCMAKE_INSTALL_PREFIX=/usr

in any case, same result more or less.
That log is complete yeah. My bad...  Ill change it to include backticks if you want.

**Comment by deathkiller** [comment]
That's strange, because it works for me in GCC/Clang/MSVC.
Can you try it again with the latest commit?

**Comment by FrostKnight** [comment]
> That's strange, because it works for me in GCC/Clang/MSVC. Can you try it again with the latest commit?

It works again. Btw, I just tried loading it and don't see any multiplayer options, is there something I am missing?

**Comment by deathkiller** [comment]
Great! You have to build it with `-D WITH_MULTIPLAYER=ON` for now. See https://deat.tk/jazz2/docs/building.html#building-config-params-game

**Comment by FrostKnight** [comment]
> Great! You have to build it with `-D WITH_MULTIPLAYER=ON` for now. See https://deat.tk/jazz2/docs/building.html#building-config-params-game

Ah i did, no one on at the moment alas though lol.

In any case, got something you should see... I used peek to record this. Its very odd indeed.

https://github.com/user-attachments/assets/ad2d6a08-4618-4417-afad-f973f2f8d623

**Comment by deathkiller** [comment]
> In any case, got something you should see... I used peek to record this. Its very odd indeed.

I just tested it and it works fine. I don't know how this can happen, maybe again something with the compiler. Do you still have GCC 8.4.0? Maybe some optimizations are not compatible with it. You can try to build it with this parameter:
`-D DEATH_USE_RUNTIME_CAST=OFF`

**Comment by FrostKnight** [comment]
> > In any case, got something you should see... I used peek to record this. Its very odd indeed.
> 
> I just tested it and it works fine. I don't know how this can happen, maybe again something with the compiler. Do you still have GCC 8.4.0? Maybe some optimizations are not compatible with it. You can try to build it with this parameter: `-D DEATH_USE_RUNTIME_CAST=OFF`

I once tried to boot steamos on a modern computer from framework and the text literally was showing up sideways. This jazz jackrabbit interaction problem is nothing on the surprise scale compared to that.

**Comment by deathkiller** [comment]
Yeah, because runtime cast is used to check object type, and according to your video, the game can't distinguish between object types - it ignores enemies and pushable objects. I don't know why it doesn't work though, maybe the compiler is just too old. I could try to install this compiler myself, but I don't think it's a big deal to leave this feature disabled.

**Comment by FrostKnight** [comment]
> Yeah, because runtime cast is used to check object type, and according to your video, the game can't distinguish between object types - it ignores enemies and pushable objects. I don't know why it doesn't work though, maybe the compiler is just too old. I could try to install this compiler myself, but I don't think it's a big deal to leave this feature disabled.

Btw, just for further information, I tried out battle and racing previews. Battle has no weapons but the standard currently, racing only has dynamite. But anywho, it appears that racing looks like it is completable with or without people. I tried it out just recently. 

Looks good so far.  I would love to be able to do battle again like my windows 95 days. Only this time, with a tank body (Linux!)

:)
